feature/enhancedSearch: Added flow implementation V-1.0.0

ponchOlvera opened the pull request:
+ Added new method in GetCharactersUseCase for Flow implementation.
+ Added UIAction and UIState flows into the viewmodel for better communication between the view and the viewmodel.
+ Added abstract class ti handle retrofit request at the use case level.
+ Added the flow collection in the SearchFragment.

ponchOlvera committed:
feature/enhancedSearch: Added flow implementation V-1.0.0

+ Added new method in GetCharactersUseCase for Flow implementation.
+ Added UIAction and UIStatte flows into the viewmodel for better communication between the view and the viewmodel.
+ Added abstract class ti handle retrofit request at the use case level.
+ Added the flow collection in the SearchFragment.

pestrada commented on app/src/main/java/com/wizeline/heroes/usecases/GetMarvelCharactersUseCase.kt:
```diff
@@ -2,19 +2,57 @@ package com.wizeline.heroes.usecases
 class GetMarvelCharactersUseCase @Inject constructor(
     @Retrofit var repository: IRepository,
-) {
+) : BaseApiResponse() {
```
Usage of flow and general code looks good. The only thing is related to Architecture.

`BaseApiResponse` belongs to the data/network/api layer. But your are making the Use Case extend from BaseApiResponse, which is like mixing Domain layer and Data layer.

Do you think we can move the BaseApiResponse to the Repository class?
That way the Use Case only interacts through the repository. The Repository is the only one who should be able to call `safeApiCall` or any other related methods.

ponchOlvera replied:
Done.

ponchOlvera committed:
Added flow implementation V-1.0.0

+ Performed code changes according to review.
+ Removed unused imports.
+ Code cleanup.
